parsers/verilog.c

```diff
@@ -1163,6 +1165,35 @@ static int skipParameterAssignment (int c)
 	return c;
 }
 
+// Using LRM Annex A Formal syntax, The declaration of function and task can be combined
+// to describe as follows (always ignore parameter_value_assignment in the class_scope):
+// --------------------------------------------------------------------------------------------------
+//                                    |  class_scope  |               |  class_scope  |
+// [ function | task ] [ lifetime ] [ { class_name :: } return_type ] { class_name :: } function_name
+//                                  |       return_type field       | |     function_name field     |
+// --------------------------------------------------------------------------------------------------
+// tokenNameSaved : Store all tokens of return_type field and function_name field
+//         In fact, any word tokens except those in `#()` will be stored, including the lifetime part
+// doubleColonPos : Record the position of `::` inserted into tokenNameSaved Array
+// return the index of first class_name in tokenNameSaved Array (return -1 if not exist)
+// NOTE: return_type field is ignored, only the class_scope of function_name field is concerned
+static int getFirstClassNameIndex (const ptrArray *const tokenNameSaved, const intArray *const doubleColonPos)
+{
+	const int tokensCount = ptrArrayCount (tokenNameSaved);
+	const int colonsCount = intArrayCount (doubleColonPos);
+	if (colonsCount == 1 && tokensCount >= 2 && intArrayItem(doubleColonPos, 0) == (tokensCount - 1))
+		return tokensCount - 2;
+	else if (colonsCount >= 2 && intArrayItem(doubleColonPos, 0) >= 1)
+	{
+		for (int i = colonsCount - 1; i >= 1; i--)
+			if (intArrayItem(doubleColonPos, i) - intArrayItem(doubleColonPos, i - 1) > 1)
+				return intArrayItem(doubleColonPos, i) - 1;
+		if (intArrayLast(doubleColonPos) == (tokensCount - 1))
+			return intArrayItem(doubleColonPos, 0) - 1;
+	}
+	return -1;
```

Comment:
> In fact, we only add the class scope to the token before ::, and :: can only appear before the return type and function name. So how to separate the return type and function name? If they both contain ::, then there are at least two adjacent tokens between the two :: that is exactly the split point. This is my implementation idea.

Here is my interpretation of your explanation.

```c
	// if there is no `::`, then no class_name is found
	if (colonsCount == 0 || intArrayItem (doubleColonPos, colonsCount - 1) != (tokensCount - 1))
		return -1;
	int startPos = intArrayItem (doubleColonPos, colonsCount - 1) - 1;
	for (int i = colonsCount - 1; i > 0; i--) {
		if (intArrayItem (doubleColonPos, i - 1) == startPos)
			// if there is no gap between two `::`, then the next class_name is found
			startPos--;
		else	// if there is a gap between two `::`, then the first class_name is found
			break;
	}
	return startPos;
```

Your version has 4 `if`s.  And mine is 2.  And same for `return`s.
How do you think?

Reply:
My original logic was to first assume that I was looking for a sequence of `name::name::...`, and then determine whether this sequence was in the scope of the *function name field*. Your logic was to first limit this sequence of `name::name::...` to the *function name field*, and then search forward. The essence is the same. It seems that your revised writing is more concise. I am preparing another PR, and then I will refactor it to your current logic, which may seem easier to understand. Then I will ping you.

Reply:
Very nice!

I was also thinking after I went to bed last night that we could use a while statement :-).